crates/astria-sequencer/src/fees/mod.rs

```diff
@@ -0,0 +1,390 @@
+
+#[async_trait::async_trait]
+impl FeeHandler for SudoAddressChange {
+    #[instrument(skip_all, err)]
+    async fn check_and_pay_fees<S: StateWrite>(&self, state: S) -> eyre::Result<()> {
+        state
+            .get_sudo_address_change_fees()
+            .await
+            .wrap_err("sudo address change fees not found, so this action is disabled")?;
+        Ok(())
+    }
+
+    fn variable_component(&self) -> u128 {
+        0
+    }
+}
+
+#[async_trait::async_trait]
+impl FeeHandler for FeeChange {
+    #[instrument(skip_all, err)]
+    async fn check_and_pay_fees<S: StateWrite>(&self, state: S) -> eyre::Result<()> {
+        state
+            .get_fee_change_fees()
+            .await
+            .wrap_err("fee change fees not found, so this action is disabled")?;
+        Ok(())
+    }
+
+    fn variable_component(&self) -> u128 {
+        0
+    }
+}
+
+#[async_trait::async_trait]
+impl FeeHandler for IbcSudoChange {
+    #[instrument(skip_all, err)]
+    async fn check_and_pay_fees<S: StateWrite>(&self, state: S) -> eyre::Result<()> {
+        state
+            .get_ibc_sudo_change_fees()
+            .await
+            .wrap_err("ibc sudo change fees not found, so this action is disabled")?;
+        Ok(())
+    }
+
+    fn variable_component(&self) -> u128 {
+        0
+    }
+}
+
+#[async_trait::async_trait]
+impl FeeHandler for IbcRelayerChange {
+    #[instrument(skip_all, err)]
+    async fn check_and_pay_fees<S: StateWrite>(&self, state: S) -> eyre::Result<()> {
+        state
+            .get_ibc_relayer_change_fees()
+            .await
+            .wrap_err("ibc relayer change fees not found, so this action is disabled")?;
+        Ok(())
+    }
+
+    fn variable_component(&self) -> u128 {
+        0
+    }
+}
+
+#[async_trait::async_trait]
+impl FeeHandler for FeeAssetChange {
+    #[instrument(skip_all, err)]
+    async fn check_and_pay_fees<S: StateWrite>(&self, state: S) -> eyre::Result<()> {
+        state
+            .get_fee_asset_change_fees()
+            .await
+            .wrap_err("fee asset change fees not found, so this action is disabled")?;
+        Ok(())
+    }
+
+    fn variable_component(&self) -> u128 {
+        0
+    }
+}
+
+#[async_trait::async_trait]
+impl FeeHandler for IbcRelay {
+    #[instrument(skip_all, err)]
+    async fn check_and_pay_fees<S: StateWrite>(&self, state: S) -> eyre::Result<()> {
+        state
+            .get_ibc_relay_fees()
+            .await
+            .wrap_err("ibc relay fees not found, so this action is disabled")?;
+        Ok(())
+    }
+
+    fn variable_component(&self) -> u128 {
+        0
+    }
+}
+
+#[instrument(skip_all, err(level = Level::WARN))]
+async fn check_and_pay_fees<S: StateWrite, T: FeeHandler + Protobuf>(
+    act: &T,
+    base: u128,
+    multiplier: u128,
+    mut state: S,
+    fee_asset: &asset::Denom,
+) -> eyre::Result<()> {
+    let total_fees = base
+        .checked_add(act.variable_component().saturating_mul(multiplier))
```

Comment:
If we do saturing arithmetic for the multiplication, I think we can also do saturating arithmetic here.